### PR TITLE
[MatchTargetSize] Extend for-loop canonicalization pattern

### PR DESCRIPTION
**Pull request by jopperm**
Support canonicalization of dependent `scf.for` loops by re-gluing individual results after the loop.

See https://github.com/intel/intel-xpu-backend-for-triton/pull/1947 for more context / the complete PoC.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp [hunk 1]
```diff
@@ -524,11 +531,25 @@ class ScfPattern : public OpRewritePattern<scf::ForOp> {
       }
 
       auto glue = cast<ttgi::GlueOp>(definingOp);
-      for (Operation *user : result.getUsers()) {
-        if (auto extract = dyn_cast<ttgi::ExtractOp>(user)) {
+      if (llvm::all_of(result.getUsers(),
+                       [](auto *user) { return isa<ttgi::ExtractOp>(user); })) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp:535]: auto* -> Operation*

File: third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp [hunk 2]
```diff
@@ -524,11 +531,25 @@ class ScfPattern : public OpRewritePattern<scf::ForOp> {
...
+        for (Operation *user : result.getUsers()) {
+          auto extract = cast<ttgi::ExtractOp>(user);
           userIndexMap[extract] = idx + extract.getIndex();
           deleteList.push_back(extract.getOperation());
         }
+      } else if (llvm::all_of(result.getUsers(), [](auto *user) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp:541]: auto * -> Operation *

**Review by victor-eds — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp [hunk 3]
```diff
@@ -514,6 +514,13 @@ class ScfPattern : public OpRewritePattern<scf::ForOp> {
 
     // Replace uses of the original loop results with the new loop results.
     userIndexMap.clear();
+
+    // If a results is used by another scf.for loop, we re-glue the individual
+    // results together to allow canonicalization of the dependent loop, too.
+    llvm::SmallDenseMap<OpResult, Value> reglueMap;
```
Comment (Dewei-Wang-sh) [inline third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp:520]: we can extend the new condition by handling the scf.yield operation.
no need to add a map.
if the scf.yield has a operand whose def is glue, then split it into two operands

**Review by Dewei-Wang-sh — CHANGES_REQUESTED** [review]

File: third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp (hunk 3 above)
Comment (jopperm) [reply third_party/intel/lib/TritonIntelGPUTransforms/MatchTargetSize.cpp]: I've simplified the code to always insert glue ops after the loop, and replacing results immediately without storing them in a map.

**Review by Dewei-Wang-sh — APPROVED** [review]
LGTM

**Pull request merged**